src/gui/tagfilterwidget.cpp

```diff
@@ -0,0 +1,225 @@
+#include "guiiconprovider.h"
+#include "tagfiltermodel.h"
+#include "tagfilterproxymodel.h"
+
+namespace
+{
+    QString getTagFilter(const TagFilterProxyModel *const model, const QModelIndex &index)
+    {
+        QString tagFilter; // Defaults to All
+        if (index.isValid()) {
+            if (!index.parent().isValid() && (index.row() == 1))
+                tagFilter = "";  // Untagged
+            else if (index.parent().isValid() || (index.row() > 1))
+                tagFilter = model->tag(index);
+        }
+
+        return tagFilter;
+    }
+}
+
+TagFilterWidget::TagFilterWidget(QWidget *parent)
+    : QTreeView(parent)
+{
+    TagFilterProxyModel *proxyModel = new TagFilterProxyModel(this);
+    proxyModel->setSortCaseSensitivity(Qt::CaseInsensitive);
+    proxyModel->setSourceModel(new TagFilterModel(this));
+    setModel(proxyModel);
+    setFrameShape(QFrame::NoFrame);
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+    setUniformRowHeights(true);
+    setHeaderHidden(true);
+    setIconSize(Utils::Misc::smallIconSize());
+#if defined(Q_OS_MAC)
+    setAttribute(Qt::WA_MacShowFocusRect, false);
+#endif
+    setContextMenuPolicy(Qt::CustomContextMenu);
+    sortByColumn(0, Qt::AscendingOrder);
+    setCurrentIndex(model()->index(0, 0));
+
+    connect(this, &TagFilterWidget::collapsed, this, &TagFilterWidget::callUpdateGeometry);
+    connect(this, &TagFilterWidget::expanded, this, &TagFilterWidget::callUpdateGeometry);
+    connect(this, &TagFilterWidget::customContextMenuRequested, this, &TagFilterWidget::showMenu);
+    connect(selectionModel(), &QItemSelectionModel::currentRowChanged, this
+            , &TagFilterWidget::onCurrentRowChanged);
+    connect(model(), &QAbstractItemModel::modelReset, this, &TagFilterWidget::callUpdateGeometry);
+}
+
+QString TagFilterWidget::currentTag() const
+{
+    QModelIndex current;
+    auto selectedRows = selectionModel()->selectedRows();
+    if (!selectedRows.isEmpty())
+        current = selectedRows.first();
+
+    return getTagFilter(static_cast<TagFilterProxyModel *>(model()), current);
+}
+
+void TagFilterWidget::onCurrentRowChanged(const QModelIndex &current, const QModelIndex &previous)
+{
+    Q_UNUSED(previous);
+
+    emit tagChanged(getTagFilter(static_cast<TagFilterProxyModel *>(model()), current));
+}
+
+void TagFilterWidget::showMenu(QPoint)
+{
+    QMenu menu(this);
+
+    QAction *addAct = menu.addAction(
+        GuiIconProvider::instance()->getIcon("list-add")
+        , tr("Add tag..."));
+    connect(addAct, &QAction::triggered, this, &TagFilterWidget::addTag);
+
+    auto selectedRows = selectionModel()->selectedRows();
+    if (!selectedRows.empty() && !TagFilterModel::isSpecialItem(selectedRows.first())) {
+        QAction *removeAct = menu.addAction(
+            GuiIconProvider::instance()->getIcon("list-remove")
```

Comment:
No, it was committed, but I just now reverted the change. When I was doing final checks of the GUI, I noticed that this had caused icons to disappear:

![tags-ui-update](https://user-images.githubusercontent.com/6699718/27380865-61603cb8-5646-11e7-9805-ba7e9b566f13.png)

This is on my test system, which is a stock Linux Mint 18.1 (no additional icon packs installed). I'm surprised, since the icon provider was supposed to fall back to the original icon if the tag-specific icon was not present, but that doesn't seem to work.